### PR TITLE
[O2-2896] When a tree was already merge, don't merge it again.

### PR DESCRIPTION
**Pull request by chiarazampolli**
This assumes that the first tree that is retrieved is always the one with the
highest cycle (--> the latest). E.g. in case of:

   KEY: TTree	O2mcparticle_001;2	O2mcparticle_001 [current cycle]
   KEY: TTree	O2mcparticle_001;1	O2mcparticle_001 [backup cycle]

this tree will be merged only 1x (otherwise it is merged on itself).

### CODE REVIEW

**Comment by ktf** [comment]
Maybe you could sort the list of trees to merge so that the later cycles come first and then filter duplicates out via std::unique (which removes all but the first occurrence in a set of consecutive duplicates).

**Comment by ktf** [comment]
Maybe you could sort the list of trees to merge so that the later cycles come first and then filter duplicates out via std::unique (which removes all but the first occurrence in a set of consecutive duplicates).

**Comment by chiarazampolli** [comment]
I don't think it is needed, my understanding is that the latest cycle is the one taken as default, which is also why it was merged 2x: the same tree was found, but the same one was taken twice.

**Comment by chiarazampolli** [comment]
@ktf , I just committed the best purging that I can think we can do - the std::unique cannot be used because we work with a TList. 

Let me know. 

As discussed at the DPG meeting, actually this merging is used at job level to merge TF's output in MC, so the problem is a blocker (even if the files I tested were all healthy).

**Comment by chiarazampolli** [comment]
Hello @ktf , @pzhristov 

Do you have further comments? I know that @jgrosseo  is away. 

Chiara

File: Common/Core/aodMerger.cxx [hunk 1]
```diff
@@ -193,14 +194,42 @@ int main(int argc, char* argv[])
...
+      }
+
       std::list<std::string> foundTrees;
 
       for (auto key2 : *treeList) {
         auto treeName = ((TObjString*)key2)->GetString().Data();
+        printf("    Processing tree %s\n", treeName);
+        bool found = (std::find(foundTrees.begin(), foundTrees.end(), treeName) != foundTrees.end());
+        if (found == true) {
+          printf("    ***WARNING*** Tree %s was already merged (even if we purged duplicated trees before, so this should not happen), skipping\n", treeName);
```
Comment (jgrosseo) [inline Common/Core/aodMerger.cxx:226]: I would propose to make a FATAL error here if this shouldn't happen

File: Common/Core/aodMerger.cxx [hunk 2]
```diff
@@ -193,14 +194,42 @@ int main(int argc, char* argv[])
...
+      for (auto i = 0; i < treeList->GetEntries(); ++i) {
+        TKey* ki = (TKey*)treeList->At(i);
+        for (int j = i + 1; j < treeList->GetEntries(); ++j) {
+          TKey* kj = (TKey*)treeList->At(j);
+          if (std::strcmp(ki->GetName(), kj->GetName()) == 0 && std::strcmp(ki->GetTitle(), kj->GetTitle()) == 0) {
+            if (ki->GetCycle() < kj->GetCycle()) {
+              printf("    ***WARNING*** we had ordered the keys, first cycle should be higher, let's continue anyway");
+            } else {
+              // key is a duplicate, let's remove it
+              treeList->Remove(kj);
```
Comment (jgrosseo) [inline Common/Core/aodMerger.cxx:210]: Given that you manipulate the list in-place while you iterate it, don't you need to add `j--` here to make sure the subsequent entry (which becomes the current on removal) is checked as well?

**Comment by jgrosseo** [comment]
Thanks for fixing this @chiarazampolli 
@ktf why do we have several cycles on a tree in O2? We write in bulk in the aod writer, so cannot this be avoided?

File: Common/Core/aodMerger.cxx (hunk 2 above)
Comment (chiarazampolli) [reply Common/Core/aodMerger.cxx:211]: Yes, of course, thanks for spotting.

File: Common/Core/aodMerger.cxx (hunk 1 above)
Comment (chiarazampolli) [reply Common/Core/aodMerger.cxx:228]: Ok, done.

**Comment by chiarazampolli** [comment]
I am recompiling O2, cannot fix right now the complaint by clang-format. 
But I actually do not understand it, I tried emacs, vi, pico, all show the lines correctly aligned, see screenshot:
image
@TimoWilken , @ktf , any idea?

**Comment by TimoWilken** [comment]
Hi @chiarazampolli, the issue isn't necessarily the alignment, it's that some of your code is indented with tab characters instead of spaces. Those are the lines clang-format is complaining about.

**Comment by jgrosseo** [comment]
Tried to fix it on github.com without success. Sorry...

**Comment by chiarazampolli** [comment]
Thanks @TimoWilken , @jgrosseo . 
In O2, we have indications of the commands to run when the tabs/spaces are not ok. O2Physics doesn't?

Chiara

**Review by jgrosseo — APPROVED** [review]

**Comment by TimoWilken** [comment]
@chiarazampolli I've added that check in #676, so hopefully it'll be available soon.

**Comment by chiarazampolli** [comment]
Thanks @TimoWilken !

**Pull request merged**